src/models/partial_tx.ts

```diff
@@ -0,0 +1,440 @@
+    // fetch token + value from txId + index
+    await txApi.getTransaction(txId, (data) => {
+      const utxo = data.tx.outputs[index];
+      const token = utxo.token_data === 0 ? HATHOR_TOKEN_CONFIG : data.tx.tokens[utxo.token_data - 1];
+      this.inputs.push(new ProposalInput(txId, index, token.uid, utxo.value, utxo.decoded.address));
+    });
+  }
+
+
+  /**
+   * Add an output to the PartialTx.
+   *
+   * @param {number} value The amount of tokens on the output.
+   * @param {Buffer} script The output script.
+   * @param {string} token The token UID.
+   * @param {Boolean} isChange If this is a change output.
+   *
+   * @memberof PartialTx
+   * @inner
+   */
+  addOutput(value: number, script: Buffer, token: string, isChange: Boolean) {
+    this.outputs.push(new ProposalOutput(
+      value,
+      script,
+      token,
+      isChange,
+    ));
+  }
+
+  /**
+   * Serialize the current PartialTx into an UTF8 string.
+   *
+   * @throws {UnsupportedScriptError} All output scripts must be P2SH or P2PKH
+   *
+   * @returns {string}
+   * @memberof PartialTx
+   * @inner
+   */
+  serialize() {
+    const changeOutputs: number[] = [];
+    this.outputs.forEach((output, index) => {
+      if (output.isChange) {
+        changeOutputs.push(index);
+      }
+    });
+    const tx = this.getTx();
+    const arr = ['PartialTx', tx.toHex(), changeOutputs.join('|')];
+    return arr.join('|');
+  }
+
+  /**
+   * Deserialize and create an instance of PartialTx
+   * This method is async because we need to fetch information on the inputs (token, value)
+   *
+   * @param {string} serialized The serialized PartialTx
+   * @param {Network} network Network used when parsing the output scripts
+   *
+   * @throws {SyntaxError} serialized argument should be valid.
+   * @throws {UnsupportedScriptError} All outputs should be P2SH or P2PKH
+   *
+   * @returns {Promise<PartialTx>}
+   * @memberof PartialTx
+   * @static
+   */
+  static async deserialize(serialized: string, network: Network) {
+    const netw = network || new Network('mainnet');
+
+    const dataArr = serialized.split('|');
+    const changeOutputs = dataArr.slice(2).map(x => +x);
+
+    if (dataArr.length < 2 || dataArr[0] !== 'PartialTx' || !changeOutputs.every(x => Number.isInteger(x))) {
+      throw new SyntaxError('Invalid PartialTx');
+    }
+
+    const txHex = dataArr[1];
+    const tx = helpers.createTxFromHex(txHex, netw);
+
+    const instance = new PartialTx(netw);
+
+    for (const input of tx.inputs) {
+      await instance.addInput(input.hash, input.index);
+    }
+
+    for (const [index, output] of tx.outputs.entries()) {
+      // validate script
+      const script = output.parseScript(netw);
+      if (!(script instanceof P2PKH || script instanceof P2SH)) {
+        throw new UnsupportedScriptError('Unsupported script type');
+      }
+
+      if (output.isAuthority()) {
+        // TODO: we dont allow passing authority on atomic swap?
```

Comment:
We don't have a limitation, but we don't have on the design any way to indicate that an input/output should be an authority.
I think we should block until we have this designed.

If we think this is required, we can change the current design to include this.

Reply:
I understand that we can't select an authority as input but I feel that having an authority in the swap should work just like any other utxo, no?

My question is, if you remove this `if`, does it work for authority? I think so, but I'm not sure.

Reply:
During the initial implementation I was having issues with `token_data` because the tokens array is not fixed during the negotiation phase so I thought we should not use it.
But we could work around this issue with extra metadata on the `ProposalOutput`, for instance, have an `authority` property with the authority mask information (`0x00` being it's not an authority).

What do you think?

Reply:
We now track `tokenData` for inputs and outputs, but the output `tokenData` is still recalculated when converting to the transaction data.
This way we can track which inputs and outputs are authorities and remove them from the balance calculation.